File: client/lib/models/page.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:client/utils/error_handling.dart';
import 'package:dio/dio.dart';

class PaginatedResponse<T> {
  final int currentPage;
  final int pageSize;
  final int totalItems;
  final int totalPages;
  final List<T> data;

  PaginatedResponse({
    required this.currentPage,
    required this.pageSize,
    required this.totalItems,
    required this.totalPages,
    required this.data,
  });

  factory PaginatedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    return PaginatedResponse(
      currentPage: json['currentPage'],
      pageSize: json['pageSize'],
      totalItems: json['totalItems'],
      totalPages: json['totalPages'],
      data: (json['data'] as List)
          .map((item) => fromJsonT(item as Map<String, dynamic>))
          .toList(),
    );
  }
  static Future<Result<PaginatedResponse<T>, String>> fetch<T>({
    required String endpoint,
    String? token,
    Duration timeout = const Duration(seconds: 10),
    Map<String, String>? queryParameters,
    required T Function(Map<String, dynamic>) fromJsonT,
  }) async {
    try {
      final dio = Dio(
        BaseOptions(
          connectTimeout: Duration(milliseconds: timeout.inMilliseconds),
          receiveTimeout: Duration(milliseconds: timeout.inMilliseconds),
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            if (token != null) 'Authorization': 'Bearer $token',
          },
        ),
      );

      final response = await dio.get(
        endpoint,
        queryParameters: queryParameters,
        options: Options(responseType: ResponseType.json),
      );

      var data = response.data;
      StatusCode statusCode = StatusCode(code: response.statusCode ?? 404);
      return statusCode.processStatus<PaginatedResponse<T>, String>(
        json: data,
        onSuccess: (json) {
          return PaginatedResponse.fromJson(
            json as Map<String, dynamic>,
            fromJsonT,
          );
        },
      );
    } catch (e) {
      return Err<PaginatedResponse<T>, String>(e.toString());
    }
  }
}

File: frontend/models/Page.ts
import { Result, Ok, Err } from "@/utils/result"; // your Result type
export interface PaginatedResponse<T> {
  currentPage: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  data: T[];
}

export function paginatedResponseFromJson<T>(
  json: any,
  fromJsonT: (item: any) => T,
): PaginatedResponse<T> {
  return {
    currentPage: json.currentPage,
    pageSize: json.pageSize,
    totalItems: json.totalItems,
    totalPages: json.totalPages,
    data: Array.isArray(json.data) ? json.data.map(fromJsonT) : [],
  };
}

export async function fetchPaginated<T>(args: {
  endpoint: string;
  token?: string;
  timeoutMs?: number;
  queryParams?: Record<string, string | number>;
  fromJsonT: (item: any) => T;
}): Promise<Result<PaginatedResponse<T>, string>> {
  const { endpoint, token, timeoutMs = 10000, queryParams, fromJsonT } = args;

  try {
    // build URL with query params, if any
    const url = new URL(endpoint);
    if (queryParams) {
      Object.entries(queryParams).forEach(([key, value]) => {
        url.searchParams.append(key, String(value));
      });
    }

    const response = (await Promise.race([
      fetch(url.toString(), {
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          ...(token ? { Authorization: `Bearer ${token}` } : {}),
        },
      }),
      new Promise<never>((_, reject) =>
        setTimeout(() => reject(new Error("Request timed out")), timeoutMs),
      ),
    ])) as Response;

    const text = await response.text();
    // check for non‑JSON errors (e.g., HTML error pages)
    if (text.trim().startsWith("<!DOCTYPE html>")) {
      return Err(
        "Server returned HTML instead of JSON — possible invalid token.",
      );
    }

    const data = JSON.parse(text);
    const paginated = paginatedResponseFromJson(data, fromJsonT);
    return Ok(paginated);
  } catch (e: any) {
    return Err(e.message ?? "Unknown error");
  }
}
